### PR TITLE
358 Convert cloud-init user-data to use a dict instead of part files

### LINKED ISSUE #358: simplify cloud-init config
Currently, cloud-init configs are written to a series of file parts and stitched together. We could, instead, just build a dict and write the yaml once. This would allow us to check the size (metadata services have maximum payload sizes that we cannot exceed) and also would be much faster.

### PR DESCRIPTION
**Pull request by joejulian**
Writing part files is not idempotent, is slow and inefficient, and requires a lot of copy/paste code which is harder to maintain.

By moving all this into a single dict, this complexity can be simplified and the use of thinks like version-templating can be done consistently. It's faster, and best of all, idempotent.

This code also provides a framework for making future changes needed for supporting other distros and providers.

Fixes #358 and #792

### CODE REVIEW

File: ansible/roles/kraken.nodePool/kraken.nodePool.master/templates/v1.5/api-server-manifest.yaml.jinja2 [hunk 1]
```diff
@@ -1,85 +0,0 @@
----
```
Comment (coffeepac) [inline ansible/roles/kraken.nodePool/kraken.nodePool.master/templates/v1.5/api-server-manifest.yaml.jinja2]: this file is still required.  kubernetes 1.5 doesn't support taints/tolerations.

**Comment by joejulian** [comment]
Rebased and fixed the missing 1.5 template as part of the rebase.

File: ansible/roles/kraken.nodePool/kraken.nodePool.master/templates/v1.5/units.kubelet.part.jinja2 [hunk 2]
```diff
@@ -4,14 +4,15 @@
...
-    Requires=kraken-ssl.service
+    Wants=docker.service
+    Wants=kraken-ssl.service
     [Service]
     EnvironmentFile=/etc/network-environment
     ExecStartPre=/usr/bin/mkdir -p /var/lib/kubelet
     ExecStartPre=/usr/bin/mkdir -p /opt/cni/bin
     ExecStartPre=/usr/bin/bash -c '/usr/bin/wget https://github.com/containernetworking/cni/releases/download/v0.3.0/cni-v0.3.0.tgz -O - | sudo tar -xz -C /opt/cni/bin'
     ExecStartPre=/usr/bin/mkdir -p /etc/cni/net.d
+    ExecStartPre=/usr/bin/systemctl --quiet is-active docker.service karken-ssl.service
```
Comment (leahnp) [inline ansible/roles/kraken.nodePool/kraken.nodePool.master/templates/v1.5/units.kubelet.part.jinja2]: "karken"

File: ansible/roles/kraken.nodePool/kraken.nodePool.node/templates/units.kubelet.part.jinja2 [hunk 3]
```diff
@@ -4,15 +4,16 @@
...
+    Wants=docker.service
+    Wants=kraken-ssl.service
     [Service]
     EnvironmentFile=/etc/network-environment
     ExecStartPre=/usr/bin/mkdir -p /var/lib/kubelet
     ExecStartPre=/usr/bin/mkdir -p /var/log/containers
     ExecStartPre=/usr/bin/mkdir -p /opt/cni/bin
     ExecStartPre=/usr/bin/bash -c '/usr/bin/wget https://github.com/containernetworking/cni/releases/download/v0.3.0/cni-v0.3.0.tgz -O - | sudo tar -xz -C /opt/cni/bin'
     ExecStartPre=/usr/bin/mkdir -p /etc/cni/net.d
+    ExecStartPre=/usr/bin/systemctl --quiet is-active docker.service karken-ssl.service
```
Comment (leahnp) [inline ansible/roles/kraken.nodePool/kraken.nodePool.node/templates/units.kubelet.part.jinja2]: karken

File: ansible/roles/kraken.nodePool/kraken.nodePool.node/templates/v1.5/units.kubelet.part.jinja2 [hunk 4]
```diff
@@ -4,15 +4,16 @@
...
+    Wants=docker.service
+    Wants=kraken-ssl.service
     [Service]
     EnvironmentFile=/etc/network-environment
     ExecStartPre=/usr/bin/mkdir -p /var/lib/kubelet
     ExecStartPre=/usr/bin/mkdir -p /var/log/containers
     ExecStartPre=/usr/bin/mkdir -p /opt/cni/bin
     ExecStartPre=/usr/bin/bash -c '/usr/bin/wget https://github.com/containernetworking/cni/releases/download/v0.3.0/cni-v0.3.0.tgz -O - | sudo tar -xz -C /opt/cni/bin'
     ExecStartPre=/usr/bin/mkdir -p /etc/cni/net.d
+    ExecStartPre=/usr/bin/systemctl --quiet is-active docker.service karken-ssl.service
```
Comment (leahnp) [inline ansible/roles/kraken.nodePool/kraken.nodePool.node/templates/v1.5/units.kubelet.part.jinja2]: karken

**Comment by coffeepac** [comment]
tested with 1.6:  works
1.5 is still broken.  will send the first of a few patches in a minute.  would you prefer I update the branch directly or send you the patch?

**Comment by joejulian** [comment]
Just the one change in Documentation/ansible/cloudconfig_snippets.md at the top of the PR where I delete the part about parts.

**Comment by joejulian** [comment]
@coffeepac Feel free.

File: ansible/roles/kraken.nodePool/kraken.nodePool.master/templates/units.kubelet.part.jinja2 [hunk 5]
```diff
@@ -4,14 +4,15 @@
     [Unit]
     After=docker.service
     After=kraken-ssl.service
-    Requires=docker.service
-    Requires=kraken-ssl.service
+    Wants=docker.service
```
Comment (davidewatson) [inline ansible/roles/kraken.nodePool/kraken.nodePool.master/templates/units.kubelet.part.jinja2]: This change gives me pause. Seems like it should be a separate PR.

**Comment by coffeepac** [comment]
v1.5 works now, had to do some things.

**Comment by joejulian** [comment]
I think that's satisfactory for 1.5 for now since this won't be a factor much longer. I have minor issues with the fix but they can wait until a similar problem appears in the future (hopefully never).

File: ansible/roles/kraken.nodePool/kraken.nodePool.master/templates/units.kubelet.part.jinja2 (hunk 5 above)
Comment (joejulian) [reply ansible/roles/kraken.nodePool/kraken.nodePool.master/templates/units.kubelet.part.jinja2]: I could do that. Wouldn't be that hard.
